### ISSUE TITLE
v1.8.0 repeatedly panics causing loss of functionality and client message cache corruption

### ISSUE THREAD

**Issue by kortschak** [ProtonMail/proton-bridge#197]
Issue tracker is ONLY used for reporting bugs with technical details. "It doesn't work" or new features should be discussed with our customer support. Please use bug report function in Bridge or contact bridge@protonmail.ch.

I allowed proton-bridge to self update from 1.6.9 to 1.8.0. Immediately on restart I noticed message header corruption and errors in scanning Drafts and Archive folders. Purging the client cache (which would normally fix the header corruption) made things worse.

## Expected Behavior

Normal behaviour, no message header corruption and all subscribed folders navigable.

## Current Behavior

The above and the following in the logs
```
ERRO[May 19 20:21:45.859] panic serving 127.0.0.1:49236: runtime error: index out of range [3] with length 0
goroutine 88 [running]:
runtime/debug.Stack(0xc000430ea0, 0x0, 0xc0005fe5a0)
	runtime/debug/stack.go:24 +0x9d
github.com/emersion/go-imap/server.(*conn).serve.func2(0xc000430ea0, 0xc0005f9f08)
	@/github.com/emersion/go-imap/server/conn.go:282 +0x191
panic(0x14ae7a0, 0xc00050a100)
	runtime/panic.go:679 +0x1b2
encoding/binary.bigEndian.Uint32(...)
	encoding/binary/binary.go:111
github.com/ProtonMail/proton-bridge/internal/store.btoi(...)
	github.com/ProtonMail/proton-bridge@/internal/store/convert.go:31
github.com/ProtonMail/proton-bridge/internal/store.(*Mailbox).txGetFinalUID(0xc000123880, 0xc000424a00, 0x0)
	github.com/ProtonMail/proton-bridge@/internal/store/mailbox_ids.go:321 +0xa9
github.com/ProtonMail/proton-bridge/internal/store.(*Mailbox).GetAPIIDsFromUIDRange.func1(0xc0002888c0, 0x0, 0xc0002888c0)
	github.com/ProtonMail/proton-bridge@/internal/store/mailbox_ids.go:50 +0x33d
go.etcd.io/bbolt.(*DB).View(0xc0001f6c00, 0xc0005f93d0, 0x0, 0x0)
	go.etcd.io/bbolt@v1.3.5/db.go:725 +0xa8
github.com/ProtonMail/proton-bridge/internal/store.(*Mailbox).GetAPIIDsFromUIDRange(0xc000123880, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	github.com/ProtonMail/proton-bridge@/internal/store/mailbox_ids.go:37 +0x93
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).apiIDsFromSeqSet(0xc0001982a0, 0xc0005f9601, 0xc0003e44c0, 0x3, 0xc0005f9838, 0x506c95, 0xba00000002, 0x506c95)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:594 +0x13e
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).listMessages(0xc0001982a0, 0x5cf06397e6201, 0xc0003e44c0, 0xc0003e44e0, 0x2, 0x2, 0xc0005fe480, 0xc0003e4500, 0x0, 0x0)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:521 +0x24f
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).ListMessages.func1(0xc02159507339eaee, 0xaa309d346)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:494 +0x68
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).logCommand(0xc0001982a0, 0xc0005f9b58, 0x157896c, 0x5, 0xc000424940, 0x4, 0x4, 0x0, 0x60)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox.go:71 +0x64
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).ListMessages(0xc0001982a0, 0x15f7401, 0xc0003e44c0, 0xc0003e44e0, 0x2, 0x2, 0xc0005fe480, 0x1452708, 0x14526f8)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:493 +0x24a
github.com/emersion/go-imap/server.(*Fetch).handle(0xc0003e44a0, 0xd66301, 0x182a7e0, 0xc000607320, 0x141ec00, 0x14526c0)
	@/github.com/emersion/go-imap/server/cmd_selected.go:175 +0x163
github.com/emersion/go-imap/server.(*Fetch).UidHandle(0xc0003e44a0, 0x182a7e0, 0xc000607320, 0x7f38e461e068, 0xc0003e44a0)
	@/github.com/emersion/go-imap/server/cmd_selected.go:200 +0x97
github.com/emersion/go-imap/server.(*Uid).Handle(0xc000286cf0, 0x182a7e0, 0xc000607320, 0xc000286cf0, 0x0)
	@/github.com/emersion/go-imap/server/cmd_selected.go:316 +0xbf
github.com/emersion/go-imap/server.(*conn).handleCommand(0xc000430ea0, 0xc0005f9ea8, 0x5, 0x8, 0x0, 0x0, 0x0)
	@/github.com/emersion/go-imap/server/conn.go:379 +0x94
github.com/emersion/go-imap/server.(*conn).serve(0xc000430ea0, 0x182a7e0, 0xc000607320, 0x0, 0x0)
	@/github.com/emersion/go-imap/server/conn.go:328 +0x7a2
github.com/emersion/go-imap/server.(*Server).serveConn(0xc00050c630, 0x182a7e0, 0xc000607320, 0x0, 0x0)
	@/github.com/emersion/go-imap/server/server.go:277 +0x121
created by github.com/emersion/go-imap/server.(*Server).Serve
	@/github.com/emersion/go-imap/server/server.go:225 +0x20e  pkg=server-imap
ERRO[May 19 20:21:45.864] panic serving 127.0.0.1:49242: runtime error: index out of range [3] with length 0
goroutine 178 [running]:
runtime/debug.Stack(0xc0003d6060, 0x0, 0xc00062a7e0)
	runtime/debug/stack.go:24 +0x9d
github.com/emersion/go-imap/server.(*conn).serve.func2(0xc0003d6060, 0xc000455f08)
	@/github.com/emersion/go-imap/server/conn.go:282 +0x191
panic(0x14ae7a0, 0xc00061a260)
	runtime/panic.go:679 +0x1b2
encoding/binary.bigEndian.Uint32(...)
	encoding/binary/binary.go:111
github.com/ProtonMail/proton-bridge/internal/store.btoi(...)
	github.com/ProtonMail/proton-bridge@/internal/store/convert.go:31
github.com/ProtonMail/proton-bridge/internal/store.(*Mailbox).txGetFinalUID(0xc000123880, 0xc0004ae700, 0x0)
	github.com/ProtonMail/proton-bridge@/internal/store/mailbox_ids.go:321 +0xa9
github.com/ProtonMail/proton-bridge/internal/store.(*Mailbox).GetAPIIDsFromUIDRange.func1(0xc00017a540, 0x0, 0xc00017a540)
	github.com/ProtonMail/proton-bridge@/internal/store/mailbox_ids.go:50 +0x33d
go.etcd.io/bbolt.(*DB).View(0xc0001f6c00, 0xc0004553d0, 0x0, 0x0)
	go.etcd.io/bbolt@v1.3.5/db.go:725 +0xa8
github.com/ProtonMail/proton-bridge/internal/store.(*Mailbox).GetAPIIDsFromUIDRange(0xc000123880, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	github.com/ProtonMail/proton-bridge@/internal/store/mailbox_ids.go:37 +0x93
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).apiIDsFromSeqSet(0xc000126700, 0xc000455601, 0xc0006066c0, 0x3, 0xc000455838, 0x506c95, 0xba00000002, 0x506c95)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:594 +0x13e
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).listMessages(0xc000126700, 0x5cf0639c2e501, 0xc0006066c0, 0xc0006066e0, 0x2, 0x2, 0xc00062a6c0, 0xc000606700, 0x0, 0x0)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:521 +0x24f
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).ListMessages.func1(0xc0215950737e6e2c, 0xaa34e566a)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:494 +0x68
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).logCommand(0xc000126700, 0xc000455b58, 0x157896c, 0x5, 0xc0004ae640, 0x4, 0x4, 0x0, 0x60)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox.go:71 +0x64
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).ListMessages(0xc000126700, 0x15f7401, 0xc0006066c0, 0xc0006066e0, 0x2, 0x2, 0xc00062a6c0, 0xc000455c38, 0xb3bc16)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:493 +0x24a
github.com/emersion/go-imap/server.(*Fetch).handle(0xc0006066a0, 0xd66301, 0x182a7e0, 0xc0003d4020, 0x141ec00, 0x14526c0)
	@/github.com/emersion/go-imap/server/cmd_selected.go:175 +0x163
github.com/emersion/go-imap/server.(*Fetch).UidHandle(0xc0006066a0, 0x182a7e0, 0xc0003d4020, 0x7f38e461e068, 0xc0006066a0)
	@/github.com/emersion/go-imap/server/cmd_selected.go:200 +0x97
github.com/emersion/go-imap/server.(*Uid).Handle(0xc000442710, 0x182a7e0, 0xc0003d4020, 0xc000442710, 0x0)
	@/github.com/emersion/go-imap/server/cmd_selected.go:316 +0xbf
github.com/emersion/go-imap/server.(*conn).handleCommand(0xc0003d6060, 0xc000455ea8, 0x5, 0x8, 0x0, 0x0, 0x0)
	@/github.com/emersion/go-imap/server/conn.go:379 +0x94
github.com/emersion/go-imap/server.(*conn).serve(0xc0003d6060, 0x182a7e0, 0xc0003d4020, 0x0, 0x0)
	@/github.com/emersion/go-imap/server/conn.go:328 +0x7a2
github.com/emersion/go-imap/server.(*Server).serveConn(0xc00050c630, 0x182a7e0, 0xc0003d4020, 0x0, 0x0)
	@/github.com/emersion/go-imap/server/server.go:277 +0x121
created by github.com/emersion/go-imap/server.(*Server).Serve
	@/github.com/emersion/go-imap/server/server.go:225 +0x20e  pkg=server-imap
ERRO[May 19 20:21:45.872] panic serving 127.0.0.1:49252: runtime error: index out of range [3] with length 0
goroutine 168 [running]:
runtime/debug.Stack(0xc0004308a0, 0x0, 0xc00062b440)
	runtime/debug/stack.go:24 +0x9d
github.com/emersion/go-imap/server.(*conn).serve.func2(0xc0004308a0, 0xc0005fdf08)
	@/github.com/emersion/go-imap/server/conn.go:282 +0x191
panic(0x14ae7a0, 0xc00061a720)
	runtime/panic.go:679 +0x1b2
encoding/binary.bigEndian.Uint32(...)
	encoding/binary/binary.go:111
github.com/ProtonMail/proton-bridge/internal/store.btoi(...)
	github.com/ProtonMail/proton-bridge@/internal/store/convert.go:31
github.com/ProtonMail/proton-bridge/internal/store.(*Mailbox).txGetFinalUID(0xc000123880, 0xc0004af500, 0x0)
	github.com/ProtonMail/proton-bridge@/internal/store/mailbox_ids.go:321 +0xa9
github.com/ProtonMail/proton-bridge/internal/store.(*Mailbox).GetAPIIDsFromUIDRange.func1(0xc00017ac40, 0x0, 0xc00017ac40)
	github.com/ProtonMail/proton-bridge@/internal/store/mailbox_ids.go:50 +0x33d
go.etcd.io/bbolt.(*DB).View(0xc0001f6c00, 0xc0005fd3d0, 0x0, 0x0)
	go.etcd.io/bbolt@v1.3.5/db.go:725 +0xa8
github.com/ProtonMail/proton-bridge/internal/store.(*Mailbox).GetAPIIDsFromUIDRange(0xc000123880, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	github.com/ProtonMail/proton-bridge@/internal/store/mailbox_ids.go:37 +0x93
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).apiIDsFromSeqSet(0xc000123ce0, 0xc0005fd601, 0xc0006076e0, 0x3, 0xc0005fd838, 0x506c95, 0xba00000002, 0x506c95)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:594 +0x13e
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).listMessages(0xc000123ce0, 0x5cf063a486b01, 0xc0006076e0, 0xc000607700, 0x2, 0x2, 0xc00062b320, 0xc000607720, 0x0, 0x0)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:521 +0x24f
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).ListMessages.func1(0xc02159507403f442, 0xaa3d3dc72)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:494 +0x68
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).logCommand(0xc000123ce0, 0xc0005fdb58, 0x157896c, 0x5, 0xc0004af440, 0x4, 0x4, 0x0, 0x60)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox.go:71 +0x64
github.com/ProtonMail/proton-bridge/internal/imap.(*imapMailbox).ListMessages(0xc000123ce0, 0x15f7401, 0xc0006076e0, 0xc000607700, 0x2, 0x2, 0xc00062b320, 0xc0005fdc38, 0xb3bc16)
	github.com/ProtonMail/proton-bridge@/internal/imap/mailbox_messages.go:493 +0x24a
github.com/emersion/go-imap/server.(*Fetch).handle(0xc0006076c0, 0xd66301, 0x182a7e0, 0xc0006068a0, 0x141ec00, 0x14526c0)
	@/github.com/emersion/go-imap/server/cmd_selected.go:175 +0x163
github.com/emersion/go-imap/server.(*Fetch).UidHandle(0xc0006076c0, 0x182a7e0, 0xc0006068a0, 0x7f38e461e068, 0xc0006076c0)
	@/github.com/emersion/go-imap/server/cmd_selected.go:200 +0x97
github.com/emersion/go-imap/server.(*Uid).Handle(0xc0004438f0, 0x182a7e0, 0xc0006068a0, 0xc0004438f0, 0x0)
	@/github.com/emersion/go-imap/server/cmd_selected.go:316 +0xbf
github.com/emersion/go-imap/server.(*conn).handleCommand(0xc0004308a0, 0xc0005fdea8, 0x5, 0x8, 0x0, 0x0, 0x0)
	@/github.com/emersion/go-imap/server/conn.go:379 +0x94
github.com/emersion/go-imap/server.(*conn).serve(0xc0004308a0, 0x182a7e0, 0xc0006068a0, 0x0, 0x0)
	@/github.com/emersion/go-imap/server/conn.go:328 +0x7a2
github.com/emersion/go-imap/server.(*Server).serveConn(0xc00050c630, 0x182a7e0, 0xc0006068a0, 0x0, 0x0)
	@/github.com/emersion/go-imap/server/server.go:277 +0x121
created by github.com/emersion/go-imap/server.(*Server).Serve
	@/github.com/emersion/go-imap/server/server.go:225 +0x20e  pkg=server-imap
```

Note that the pretty colours that are included in the logs make log handling very much harder when using a text editor. Please don't do that.

## Possible Solution

The error appears to be due to failing to check whether `"github.com/ProtonMail/proton-bridge/internal/store".(*MaiBox).txGetFinalUID` gets a valid UID from Bolt and then handing the invalid (probably nil, though the stack trace does not show this) slice on to `binary.BigEndian.Uint32` which prompty reaches out of bounds.

Note that the docs for the `Last` method on `*Cursor` say, 

> Last moves the cursor to the last item in the bucket and returns its key and value.
> **If the bucket is empty then a nil key and value are returned.**
> The returned key and value are only valid for the life of the transaction.

This should probably be checked.

## Steps to Reproduce

1. Allow auto-update to v1.8.0.
2. Watch the world burn.

## Version Information

v1.8.0

## Context (Environment)

n/a

## Detailed Description

n/a

## Possible Implementation

n/a

**Comment by andrzejsza** [comment]
Thank you for raising this. We overlooked the case where a client does a UID FETCH of a range on an empty mailbox. A hotfix is coming.

**Comment by andrzejsza** [comment]
fixed in 1.8.1.
